src/changelog/prepareChangelogPaths.ts

```diff
@@ -0,0 +1,109 @@
+ */
+export function prepareChangelogPaths(params: {
+  options: Pick<BeachballOptions, 'changelog' | 'generateChangelog'>;
+  packageName: string;
+  /** cwd where changelogs are located */
+  changelogAbsDir: string;
+}): ChangelogPaths {
+  const { options, changelogAbsDir: cwd, packageName } = params;
+  const { changelog, generateChangelog } = options;
+  const { uniqueFilenames } = changelog || {};
+
+  const paths: ChangelogPaths = {};
+
+  if (!generateChangelog) {
+    return paths;
+  }
+
+  for (const ext of ['md', 'json'] as const) {
+    if (generateChangelog !== true && generateChangelog !== ext) {
+      continue; // not generating this file type
+    }
+
+    const defaultPath = path.join(cwd, `CHANGELOG.${ext}`);
+
+    // Generate a unique filename based on the package name hash.
+    const hash = crypto.createHash('md5').update(packageName).digest('hex').slice(0, 8);
+    const pathWithHash = path.join(cwd, `CHANGELOG-${hash}.${ext}`);
+
+    // Choose the changelog file path based on whether unique filenames are enabled.
+    const filePath = (paths[ext] = uniqueFilenames ? pathWithHash : defaultPath);
+
+    if (!fs.existsSync(filePath)) {
+      // The path to use doesn't exist--check if the other path does, and if so, rename it.
+      const renamed = renameIfPresent({
+        oldPath: uniqueFilenames ? defaultPath : pathWithHash,
+        newPath: filePath,
+      });
+
+      if (!renamed) {
+        // Neither path exists. Check for other hashed changelog files in case the package was renamed.
+        // (If there's no other hashed file, it's the first changelog entry for this package.)
+        const existingHashedPath = getExistingHashedPath({ cwd, ext });
+        if (existingHashedPath) {
+          renameIfPresent({ oldPath: existingHashedPath, newPath: filePath });
+        }
+      }
+    }
+  }
+
+  return paths;
+}
+
+/**
+ * Get the path to the newest `CHANGELOG-{hash}.{ext}` file in the given directory.
+ */
+function getExistingHashedPath(params: { cwd: string; ext: 'md' | 'json' }): string | undefined {
+  const { cwd, ext } = params;
+  const fileRegexp = new RegExp(`^CHANGELOG-[a-f\\d]{8}\\.${ext}$`);
+  try {
+    let newestDate = 0;
+    let newestFile: string | undefined;
+
+    for (const file of fs.readdirSync(cwd)) {
+      if (!fileRegexp.test(file)) {
+        continue;
+      }
+
+      const date = fs.statSync(path.join(cwd, file)).mtimeMs;
+      if (date > newestDate) {
+        newestDate = date;
+        newestFile = file;
+      }
+    }
+    return newestFile ? path.join(cwd, newestFile) : undefined;
+  } catch (e) {
+    console.warn(`Error getting changelog file info in ${cwd}: ${e}`);
+  }
+}
+
+function renameIfPresent(params: { oldPath: string; newPath: string }): boolean {
+  const { oldPath, newPath } = params;
+  try {
+    if (fs.existsSync(oldPath)) {
+      fs.renameSync(oldPath, newPath);
```

Comment:
Existence checking files before operating on them is generally an anti-pattern that is advised against by the authors of the file system APIs, since in theory the existence can change between the calls, so you have to handle the non-existence during the real call anyway. Simpler to just handle the `ENOENT` or `ENOTDIR` during the `renameSync` call.